wgsl/index.bs

```diff
@@ -7272,13 +7270,337 @@ See [[#expressions]].
+- Create two new nodes, named "Result" and "CF_after"
+- If the control-flow at the start of the function was tagged AlwaysMustBeUniform, then add an edge from MustBeUniform to the last CF_i
+- Otherwise add an edge from CF_after to the last CF_i
+- If the function was tagged ControlFlowCannotBeUniformAfter, then add an edge from CF_after to CannotBeUniform
+- Else if it was tagged ReturnValueCannotBeUniform, then add an edge from Result to CannotBeUniform
+- Add an edge from Result to CF_after
+- For each argument *i*:
+    - If the corresponding parameter was tagged AlwaysMustBeUniform, then add an edge from MustBeUniform to arg_i
+    - Else if it was tagged MustBeUniformForControlFlowAfter, then add an edge from CF_after to arg_i
+    - Else if it was tagged MustBeUniformForReturnValue, then add an edge from Result to arg_i
+
+Note: Notice that this rule only requires adding a number of edges bounded by 3 + the number of parameters of the functions, independently of how complex the implementation of the function might be. This is key to the linear complexity of the overall algorithm.
+
+Most built-in functions have tags of:
+- for the control-flow in which they can be called: MustBeUniformForControlFlowAfter
+- for the function as a whole: NoRestriction
+- for each parameter: MustBeUniformForReturnValue
+
+Here is the list of exceptions:
+- All functions in [[#sync-builtin-functions]] are AlwaysMustBeUniform
+- All functions in [[#derivative-builtin-functions]], [[#texturesample]], [[#texturesamplebias]], and [[#texturesamplecompare]] are AlwaysMustBeUniform and ReturnValueCannotBeUniform
+
+### Uniformity rules for expressions ### {#uniformity-expressions}
+
+The rules for analyzing expressions take as argument both the expression itself and the node corresponding to control flow at the beginning of it (which we'll note "CF" below) and return the following:
+
+* A node corresponding to control flow at the exit of it
+* A node corresponding to its value
+* A set of new nodes and edges to add to the graph
+
+<table class='data'>
+  <caption>Uniformity rules for expressions (in normal rvalue position)</caption>
+  <thead>
+    <tr><th>Expression<th>New nodes<th>Recursive analyses<th>Resulting control flow node, value node<th>New edges
+  </thead>
+  <tr><td class="nowrap">*e1* || *e2*
+      <td rowspan=2>
+      <td rowspan=2 class="nowrap">(*CF*, *e1*) => (*CF1*, *V1*)<br>
+          (*V1*, *e2*) => (*CF2*, *V2*)
+      <td rowspan=2 class="nowrap">*CF2*, *V2*
+      <td rowspan=2>
+  <tr><td class="nowrap">*e1* && *e2*
+  <tr><td class="nowrap">Literal
+      <td>
+      <td>
+      <td class="nowrap">*CF*, *CF*
+      <td>
+   <tr><td class="nowrap">reference to function-scope variable or parameter "x"
```

Comment:
The spec distinguishes between variables and let-declarations, so to be precise we should name let-declarations here.
```suggestion
  <tr><td class="nowrap">reference to function-scope variable, let-declaration or formal parameter "x"
```

Separately are these rules priority based? This rule conflicts with builtin value rules below since they may be function parameters (or members of a struct function parameter).